### ISSUE TITLE
Removing a header from request in a filter is not affecting response.

### ISSUE THREAD

**Issue by RaviH** [micronaut-projects/micronaut-core#927]
My goal is to remove `Accept-Encoding` header for a particular URL. So I added a filter and removed the header in the filter before doing `chain.proceed(request)` but even though request has the header removed, `micronaut` still compresses the response.

`curl -H "Accept-Encoding: gzip" "http://localhost:8080/lineupcacheservice/simple/hello"` and it will compress the response even though in the logs following is printed:

```
19:26:39.595 [nioEventLoopGroup-1-2] INFO  m.swagger.sample.SwaggerFilter - accept-encoding header: gzip
19:26:39.607 [nioEventLoopGroup-1-2] INFO  m.swagger.sample.SwaggerFilter - accept-encoding header: null
```

[micronaut-swagger-sample.zip](https://github.com/micronaut-projects/micronaut-core/files/2593668/micronaut-swagger-sample.zip)

**Comment by graemerocher** [comment]
Don't believe you can do what you are doing since the request and headers are not mutable. 

One thing you can do is add a bean that is a subclass of `ChannelInitializer` in Netty and in the `initChannel` method remove the `http-compressor` handler which will disable compression.

We can probably add a configuration option for the future to allow disabling compression.

**Comment by RaviH** [comment]
@jameskleeh The use case is the QA would hit the swagger URL in a browser to develop there test cases, but Zuul is adding `Accept-Encoding` header, `microonaut` is honoring it (correctly I might add). But, if the content is compressed than they cannot view it in a browser.

@graemerocher 

> One thing you can do is add a bean that is a subclass of ChannelInitializer in Netty and in the initChannel method remove the http-compressor handler which will disable compression.

I am trying it, no luck so far. I am even trying to do `@Replaces(HttpContentCompressor)` with my own class but for some reason it's not taking it.

I will keep plugging away, but being able to selectively disable compression would be a great feature. I want compression except for the swagger ui html endpoint. Something like `@DisableCompression` on the rest endpoint / class would be awesome.

**Comment by jameskleeh** [comment]
@RaviH Browsers definitely know how to decompress content. What scenario is causing the browser to not decompress the content ?

**Comment by RaviH** [comment]
@jameskleeh I am not sure what I am missing:

I don't see the same behaviour: 
![image](https://user-images.githubusercontent.com/141136/48727247-90413c80-ebe6-11e8-9225-d28a1dc802e9.png)

![image](https://user-images.githubusercontent.com/141136/48727346-d1d1e780-ebe6-11e8-9ece-24d54d5b9826.png)

**Comment by jameskleeh** [comment]
@RaviH What is the value of the `Content-Encoding` header in the response?

**Comment by RaviH** [comment]
@jameskleeh `< content-encoding: gzip`

**Comment by RaviH** [comment]
@jameskleeh Scratch that. Zuul seems to be removing the `content-encoding` header:

```
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 172.28.166.177...
* TCP_NODELAY set
* Connected to specci.partnerapi.engprod-charter.net (172.28.166.177) port 80 (#0)
> GET /api/pub/lineupcacheservice/swagger-ui.html HTTP/1.1
> Host: specci.partnerapi.engprod-charter.net
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: text/html;charset=UTF-8
< Date: Mon, 19 Nov 2018 18:41:07 GMT
< Server: Apache-Coyote/1.1
< X-Originating-URL: http://specci.partnerapi.engprod-charter.net/api/pub/lineupcacheservice/swagger-ui.html
< X-Zuul: zuul
< X-Zuul-instance: unknown
< Content-Length: 503
< Connection: keep-alive
<
```

**Comment by jameskleeh** [comment]
Then I would figure out why Zuul is doing that and prevent it from doing so

**Comment by RaviH** [comment]
@jameskleeh @graemerocher  My hands are tied. Zuul team is swamped and have been very slow to respond. I have to modify my code around their response. 

In lieu of that, and in general framework flexibility/extensibility, it would be nice to be able to any/all of the below:

a. Disable compression for the app/service.
b. Disable compression for a given endpoint.
c. Allow modification of request headers. I know it's frowned upon but there are extreme cases where this may become necessary, especially when working with legacy apps, outside teams etc.

What do you think?

**Comment by jameskleeh** [comment]
The only one that seems reasonable to me is "c"

**Comment by RaviH** [comment]
@graemerocher mentioned in the Gitter room that he wouldn't mind doing (a). 

(c) Obviously is the most powerful / provides most flexibility in a myriad of scenarios, not only mine.

(a) seems more clean.

If I had to vote, I would vote on (c). But whatever micronaut team deems most suitable / correct.

In lieu of that should I open a different ticket for (c)?

**Comment by RaviH** [comment]
@graemerocher @jameskleeh Should I create separate tickets for (a), (b), (c) and close this one or keep this open and let micronaut team decide which route it takes?

**Comment by graemerocher** [comment]
my personal vote is to simply add a configuration option to enable/disable compression (option A). We should not be changing the mutability of the request (besides that is probably already possible with a request wrapper)

**Comment by jameskleeh** [comment]
Making the compression threshold configurable seems like the best idea to me. I can see that being beneficial for users that want to tune that for their environment. It also provides the ability to effectively disable compression if its set high enough

**Comment by RaviH** [comment]
Hi @graemerocher,

I tried wrapping the request with `HttpRequestWrapper` and couple other ways, but that request is pretty immutable :) . So, I don't think (and I could be wrong) there's any way to mutate the request unless we do (c).

So my vote is firmer on (c) because request headers mutability would be awesome. That would even allow adding headers and not just removing and would be useful beyond just my problem of compression (which I have solved by modifying `NettyHttpServer`). My 2 cents.
